### ISSUE TITLE
CME on server shutdown

### ISSUE THREAD

**Issue by DeathrushW** [Xephi/AuthMeReloaded#992]
http://hastebin.com/otetuvonow.lua

When the server closes, this error appears
sry bad english

**Comment by sgdc3** [comment]
Not related to CombatTagPlus

**Comment by sgdc3** [comment]
@ljacqu any idea?

**Comment by ljacqu** [comment]
First thing to say about the stacktrace is that Bukkit defines its own class loader. This is why you'll see the stacktrace going from Hikari to Bukkit classes (reading from the bottom up), which might be a little surprising at first sight... :smile:

```
Exception in thread "AuthMe-DataSource#close"
[CombatTagPlus] Disabling CombatTagPlus v1.2.3
java.util.ConcurrentModificationException
       at java.util.LinkedHashMap$LinkedHashIterator.nextNode(Unknown Source)
       at java.util.LinkedHashMap$LinkedKeyIterator.next(Unknown Source)
       at org.bukkit.plugin.java.JavaPluginLoader.getClassByName(JavaPluginLoader.java:189)
       at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:98)
       at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:87)
       at java.lang.ClassLoader.loadClass(Unknown Source)
       at java.lang.ClassLoader.loadClass(Unknown Source)
       at fr.xephi.authme.libs.zaxxer.hikari.pool.HikariPool.closeConnection(HikariPool.java:411)
       at fr.xephi.authme.libs.zaxxer.hikari.pool.HikariPool.softEvictConnection(HikariPool.java:521)
       at fr.xephi.authme.libs.zaxxer.hikari.pool.HikariPool.softEvictConnections(HikariPool.java:338)
       at fr.xephi.authme.libs.zaxxer.hikari.pool.HikariPool.shutdown(HikariPool.java:207)
       at fr.xephi.authme.libs.zaxxer.hikari.HikariDataSource.close(HikariDataSource.java:304)
       at fr.xephi.authme.datasource.MySQL.close(MySQL.java:674)
       at fr.xephi.authme.datasource.CacheDataSource.close(CacheDataSource.java:166)
       at fr.xephi.authme.initialization.TaskCloser.run(TaskCloser.java:73)
       at java.lang.Thread.run(Unknown Source)
```

Now, if we look at where the error occurred (excerpt of decompiled Bukkit code):
```java
    Class<?> getClassByName(String name) {
            // [...]
            Iterator var4 = this.loaders.iterator();
            while(var4.hasNext()) {
```

`this.loaders` was changed while we were in that `while` loop. Every plugin has its own loader, so some plugin was disabled and its loader was removed from the list. We can encounter this situation because we schedule the TaskCloser thread and Bukkit might go and disable other plugins in the meantime.

TL;DR when AuthMe gets disabled we schedule a thread to close any open tasks. This interferes with Bukkit's process of loading/unloading plugins. If this is a rare occurrence I would suggest not to try to handle this.

**Comment by sgdc3** [comment]
We should lock the main thread while disabling authme, does it make sense? @ljacqu

**Comment by ljacqu** [comment]
Time to call @games647 :P
He created the task closing logic in 5b2909d7d69e56ad5c429a99cd9cbb79d415bfc7, and there it was in the main thread. @DNx5 later on removed it and added it back in ed0ed2faf7320abc49a4d51b227f8840de551255, this time in a new thread.

Sadly none of those commits have an issue number so I can't really find any more context and I don't remember the specifics.

**Comment by games647** [comment]
The reason behind my change is that the main thread should wait until all async tasks are finished. This means that the server process should not be closed until all pending save actions are finished. 

This thing was added, because we experienced that some async tasks like AsyncQuit still needed to schedule some SQL queries when plugin will be disabled and we already closed the executor service for future tasks. (This is what I have still in mind. Correct me if I'm wrong)

Thanks to @ljacqu great explanation (https://github.com/Xephi/AuthMeReloaded/issues/992#issuecomment-257146850). It shows that this wait code block shouldn't be run async, because it's not thread-safe on shutdown.

BTW: the metrics task is async and it will wait for that one too.

I don't know if this code is really still necessary and I have to admit that I was very confused by the difference between getActiveWorkers and getPendingTasks. 

Next time I should really add a comment section about why this code exists, especially because this code isn't so clear.

**Comment by ljacqu** [comment]
Thanks for the detailed answer, @games647. What do you suggest we do? I reckon if we do it properly it should be fine to do in the main thread. Processing pending tasks should not really take that long. 

I remember a while back when the code was in the main thread people were waiting for a really long time for AuthMe to shut down; I don't remember if we ever found the cause for that.

**Comment by games647** [comment]
@ljacqu I changed it to a total wait time and not to a individual one. So it should at the current state wait max 1 min in total (+ some ms for checking how many tasks are still running). So I guess it would be fine to just let it run on the main thread again.

**Comment by sgdc3** [comment]
It should be fixed now ;)
